src/github/github.service.ts

```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  async saveToken(userId: number, accessToken: string) {
+    const user = await this.userRepository.findOne({
+      where: { githubUserId: `${userId}` },
+    });
+    if (!user) {
+      throw new HttpException(ApiMessages.NOT_FOUND, HttpStatus.NOT_FOUND);
+    }
+    user.accessToken = accessToken;
+    await this.userRepository.save(user);
+  }
+
+  async findUserById(userId: string) {
+    return await this.userRepository.findOne({
+      where: { githubUserId: userId },
+    });
+  }
+
+  async handleInstallation(req: TRequest, installationId: string) {
+    try {
+      const userId = req.user['userId'];
+      const user = await this.userRepository.findOne({
+        where: { githubUserId: userId },
+      });
+
+      if (!user) {
+        throw new HttpException('User not found', HttpStatus.NOT_FOUND);
+      }
+      await this.validateAndUpdateInstallations(user);
+      const existingInstallation =
+        await this.userInstallationRepository.findOne({
+          where: { installationId: installationId },
+        });
+
+      if (existingInstallation) {
+        return { message: 'Installation ID up-to-date' };
+      }
+
+      const installationToken =
+        await this.generateInstallationAccessToken(installationId);
+      const repositories =
+        await this.fetchOrganizationDetails(installationToken);
+
+      if (!repositories || repositories.length === 0) {
+        throw new HttpException('No repositories found', HttpStatus.NOT_FOUND);
+      }
+
+      const organizationName = repositories[0].owner.login;
+
+      let userInstallation = await this.userInstallationRepository.findOne({
+        where: { user: { id: user.id }, organizationName: organizationName },
+      });
+
+      if (!userInstallation) {
+        userInstallation = new UserInstallation();
+        userInstallation.organizationName = organizationName;
+        userInstallation.installationId = installationId;
+        userInstallation.user = user;
+      } else if (userInstallation.installationId !== installationId) {
+        userInstallation.installationId = installationId;
+      } else {
+        return { message: 'Installation ID already up-to-date' };
+      }
+
+      await this.userInstallationRepository.save(userInstallation);
+      return { message: 'Installation ID saved/updated successfully' };
+    } catch (error) {
+      if (error.status === '404') {
+        throw new NotFoundException('App not installed');
+      }
+      throw new HttpException(
+        `Failed to handle installation: ${error.message}`,
+        HttpStatus.BAD_REQUEST,
+      );
+    }
+  }
+
+  private hasDuplicates(items: number[]): boolean {
+    for (let i = 0; i < items.length; i++) {
+      for (let j = i + 1; j < items.length; j++) {
+        if (items[i] === items[j]) {
+          console.log('items[i]', items[i], 'items[j]', items[j]);
+          return true;
+        }
+      }
+    }
+    return false;
+  }
```

Comment:
PullProofAI

- The `hasDuplicates` method can be optimized for performance by using a Set to check for duplicates.

```diff
+ private hasDuplicates(items: number[]): boolean {
+     const set = new Set();
+     for (const item of items) {
+         if (set.has(item)) {
+             console.log('Duplicate item found:', item);
+             return true;
+         }
+         set.add(item);
+     }
+     return false;
+ }
```

### Summary
The code changes look good overall. By improving error handling with more specific messages and optimizing the `hasDuplicates` method, the code will be more robust and maintainable.


<!-- This is an auto-generated comment by PullProofAI -->